packages/core/src/StateMachine.ts

```diff
@@ -252,10 +249,18 @@ export class StateMachine<
    * @param event The received event
    */
   public transition(
-    state: StateValue | State<TContext, TEvent, TResolvedTypesMeta> | undefined,
-    event: Event<TEvent> | SCXML.Event<TEvent>
+    state: StateValue | State<TContext, TEvent, TResolvedTypesMeta> = this
+      .initialState,
+    event: Event<TEvent> | SCXML.Event<TEvent>,
+    actorCtx?: ActorContext<TEvent, State<TContext, TEvent, any>>
   ): State<TContext, TEvent, TResolvedTypesMeta> {
-    return this._transition(state, event);
+    const currentState = toState(state, this);
+    const scxmlEvent = toSCXMLEvent(event);
+
+    const nextState = macrostep(currentState, scxmlEvent, this, actorCtx);
+    nextState._sessionid = actorCtx?.sessionId ?? currentState._sessionid;
```

Comment:
q: why this is not handled when creating the new State somewhere inside the `macrostep`?

Reply:
It should be; `v5/simplify-algorithm` needs to be merged in first